feat: Adding GoReleaser to handle releases

sc250024 opened the pull request:
Enables this repository to use [GoReleaser](https://github.com/goreleaser/goreleaser) to handle all releases. This PR is **_basically_** ready, it just needs some feedback from the maintainers.

## Description

By merging this PR, this repository will use [GoReleaser](https://github.com/goreleaser/goreleaser) for releasing of all artifacts. This includes, but is not limited to:

- Docker images
- GitHub releases with Go binaries pre-packaged
- Homebrew formulas

There are a number of issues to be addressed / questions to be answered by the maintainers of this project in order to merge this PR. See **Generic questions**.

The following is an example of the `_output` directory after running `make cross`, which triggers `goreleaser` ultimately:

```shell
$ tree _output

_output
├── config.yaml
├── goreleaserdocker799593599
│   ├── Dockerfile
│   └── jb
├── jb-darwin-amd64.tar.gz
├── jb-linux-amd64.tar.gz
├── jb-linux-arm64.tar.gz
├── jb-linux-armv6.tar.gz
├── jb-linux-armv7.tar.gz
├── jb-windows-amd64.zip
├── jb.rb
├── jb_darwin_amd64
│   └── jb
├── jb_linux_amd64
│   └── jb
├── jb_linux_arm64
│   └── jb
├── jb_linux_arm_6
│   └── jb
├── jb_linux_arm_7
│   └── jb
├── jb_v0.4.0-SNAPSHOT-7a629bf_checksums.txt
└── jb_windows_amd64
    └── jb.exe

7 directories, 17 files
```

## Motivation and Context

This fixes issue https://github.com/jsonnet-bundler/jsonnet-bundler/issues/64, which has been open for a while. Essentially, it enables `jsonnet-bundler` to be packaged more easily so it can be distributed amongst the main channels, i.e. Docker, Homebrew, etc. 

## Items to Address

### Generic questions

1. Should GoReleaser be triggered automatically by Drone on a `tag` push, or should a maintainer perform the releasing manually?
2. Are Docker images wanted? See **Docker registry** below.
3. Is a Homebrew formula wanted? See **Homebrew formula** below.
4. What artifacts should be attached to the GitHub release? Binaries only? Archives only? Both? See **GitHub releases**.

### Docker registry

In order for this to be merged, the repository maintainers should decide _**where**_ to put Docker images, if at all. If it's desired to not have Docker images generated, that's also possible. If it's decided that this is wanted, credentials would also need to be generated to push to the Docker registry.

Essentially, the GoReleaser builds tags according to semver specifications. For example, if this project is hosted on DockerHub, the following images would be generated:

- `docker.io/jsonnet-bundler/jb:latest`
- `docker.io/jsonnet-bundler/jb:v0.4.0`
- `docker.io/jsonnet-bundler/jb:v0.4`
- `docker.io/jsonnet-bundler/jb:v0`

Other options are also possible.

### Homebrew formula

In a similar fashion, if it's decided to create Homebrew formulas, the following options are available:

- **Option 1:** Host the Homebrew tap from this repository, which means creating a `Formula` folder at the root.
- **Option 2:** Host the Homebrew tap from another repository in the `jsonnet-bunlder` organization; in which case, a Github token would need to be provided by the maintainers which has access to write to that repository.
- **Option 3:** Don't host any Homebrew tap, and later manually create a formula in the https://github.com/Homebrew/homebrew-core repository.

### GitHub releases

Currently the `jb` binary is added to GitHub releases directly without packaging. This is also possible with GoReleaser in the exact same way, but **if Homebrew is being used as well**, then Homebrew _**needs**_ archives to be built. In other words, Homebrew does not download binaries directly; it always first downloads an archive, and then extracts it.

Coupled with the Homebrew options above are the following options for GitHub releases:

- **Option 1:** Don't use Homebrew at all; attach binaries to the GitHub release as it happens right now.
- **Option 2:** Use Homebrew in some way; attach archives only to the GitHub release.
- **Option 3:** Use Homebrew in some way; attach both archives and binaries to the GitHub release, and let people choose.

## How Has This Been Tested?

To test locally, you can either run `goreleaser --rm-dist --skip-publish --snapshot` or `make cross`; they do the same thing, in the end.

Someone who has write access to this repository should probably do more testing related to releases in order to make sure it works.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

sc250024 commented:
/ping @metalmatze @sh0rez

sh0rez commented:
Hi @sc250024, thanks a lot for setting this up!

Let me answer your questions from my personal perspective. This is not authoritative for all maintainers, but it's what I consider good and hopefully aligns with the others:

> Should GoReleaser be triggered automatically by Drone on a tag push, or should a maintainer perform the releasing manually?

Yes! All artifacts should be made in a reproducible by the CI. The GH release should only be a Draft and require manual publishing.

> Are Docker images wanted? See Docker registry below.

Yes! Having docker images allows other Dockerfiles to use `FROM jb; COPY /usr/local/bin/jb .`, which is real handy.

> Is a Homebrew formula wanted? See Homebrew formula below.

Yes!

> What artifacts should be attached to the GitHub release? Binaries only? Archives only? Both? See GitHub releases.

`jb` is fairly small, I'd vote for plain binaries as it requires least tooling on the consumer side

---

> ### Docker registry

Not sure whether it's actually possible to use goreleaser for building the docker images – Drone CI is fairly restrictive with root permissions, you can't just run a docker daemon.
You need to use their `docker` plugin, which is whitelisted server-side to elevate it's permissions.

Unless I miss something, we need to keep this in mind :)

> ### Homebrew formula

I'd really favor option 3, having it in `homebrew-core`. Updating there is fairly straight-forward and it would be provided to every user as-is, without needing taps.

> ### GitHub releases

Plain binaries should work just fine. We do the same for [Tanka](https://github.com/grafana/tanka/releases/tag/v0.11.1), and we have a working brew Formula as well: https://formulae.brew.sh/formula/tanka

---

Again, thanks for taking the time to look at this. This is super cool!!

sh0rez commented on .drone.jsonnet:
```diff
@@ -30,6 +30,14 @@
     ],
   },
 
+  local release(version) = std.mergePatch(
+    golang(version) { commands: ['curl -sL https://git.io/goreleaser | bash'] },
+    ({
+      environment: { GITHUB_TOKEN: { from_secret: 'github_token' } },
+      when: { event: 'tag' },
+    })
+  ),
```
`std.mergePatch` is not required to be used manually in Jsonnet. Instead, you can use syntax features:

```jsonnet
  local release(version) = golang(version) {
    commands: ['curl -sL https://git.io/goreleaser | bash'] },
    environment+: { GITHUB_TOKEN: { from_secret: 'github_token' } },
    when: { event: 'tag' },
  },
```

Notice the use of `+:` (merge) and `:` (override)

sc250024 replied:
Ah damn, I always wondered what the `+:` syntax meant. I will fix in next commit.

sh0rez replied:
https://tanka.dev/jsonnet/overview#merging

sc250024 replied:
@sh0rez Nice! Pushed just now.

sh0rez commented on .goreleaser.yml:
```diff
@@ -0,0 +1,58 @@
+    - go mod tidy
+    - go generate ./...
+brews:
+  - description: The jsonnet-bundler is a package manager for Jsonnet.
```
```suggestion
  - description: jsonnet-bundler is a package manager for Jsonnet.
```

I think we usually refer to it without a "the"

sc250024 replied:
This section will go away entirely since there's no self-hosting of the Homebrew formula.

sc250024 replied:
Pushed just now.

sc250024 commented:
@sh0rez Perfect. I will have to update the `.goreleaser.yml` file based on the changes you mentioned, assuming everyone else doesn't have further comments.

To reply to some of the items:

> Yes! All artifacts should be made in a reproducible by the CI. The GH release should only be a Draft and require manual publishing.

GoReleaser supports this.

> Not sure whether it's actually possible to use goreleaser for building the docker images.

It can. See here: https://github.com/sc250024/jsonnet-bundler/blob/feat-GoReleaser/.goreleaser.yml#L41-L51

> I'd really favor option 3, having it in homebrew-core. Updating there is fairly straight-forward and it would be provided to every user as-is, without needing taps.

Alright, that means we don't need a `brews` section here, and therefore, Tar / ZIP archives are not needed then. I'll make those changes.

sh0rez commented on .goreleaser.yml:
```diff
@@ -0,0 +1,58 @@
+      - 6
+      - 7
+    goos:
+      - darwin
+      - linux
+      - windows
```
Will that also attempt windows/arm builds, etc? I think we only need:

```
linux/amd64
linux/arm64
linux/arm
darwin/amd64
windows/amd64
```

and due to apple silicon maybe soon `darwin/arm64`

sc250024 replied:
For now, it's only generating the ones you mentioned. If you look at the `tree` output above in the PR comments, it generates the following archives:

```
jb-darwin-amd64.tar.gz
jb-linux-amd64.tar.gz
jb-linux-arm64.tar.gz
jb-linux-armv6.tar.gz
jb-linux-armv7.tar.gz
jb-windows-amd64.zip
```

sc250024 replied:
Pushed just now.

sh0rez commented on Dockerfile:
```diff
@@ -1,5 +1,5 @@
-FROM busybox:1.28
+FROM busybox:1.32
```
If nobody objects, let's use alpine here

sc250024 replied:
Will change in the next commit.

sh0rez commented on go.mod:
```diff
@@ -5,7 +5,6 @@ go 1.12
 require (
 	github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc // indirect
 	github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf // indirect
-	github.com/campoy/embedmd v1.0.0 // indirect
```
go.mod setup is a bit borked here, this is required for generating the markdown file (go generate)

sc250024 replied:
These lines are removed because of the `before.hooks` sections in the `.goreleaser.yml`:

```yaml
before:
  hooks:
    - go mod tidy
    - go generate ./...
```

I dug into this a bit, and I think they're still not needed. You make a good point that `embedmd` is needed for generating documentation, but if you look at the `Makefile`, it actually tries to avoid putting that library in the `go.mod` file:

```makefile
embedmd:
	pushd /tmp && GO111MODULES=on go get github.com/campoy/embedmd && popd
```

As far as I know, using that trick basically avoids putting it into the `go.mod` and `go.sum` files, right?

sh0rez replied:
Oh right, I changed that line some time ago. Let's go ahead, it shouldn't break. Fingers crossed

sh0rez commented:
> It can. See here: https://github.com/sc250024/jsonnet-bundler/blob/feat-GoReleaser/.goreleaser.yml#L41-L51.

Only if it has a docker daemon in CI, right? And I'm not quite sure you can get one without the plugin

> Where should the Docker repo be hosted?

Given I got https://hub.docker.com/orgs/jsonnet, `docker.io/jsonnet/bundler` sounds neat to me

brancz commented:
No strong opinion on all of this, obviously +1 on automation, all I want to make sure and emphasize is I want to ensure that `go get` will always continue to work.

> Given I got https://hub.docker.com/orgs/jsonnet, docker.io/jsonnet/bundler sounds neat to me

That makes this look more official than it currently is, I think it should be `jsonnet-bundler/jsonnet-bundler`. As I said I don't feel too strongly, but docker containers don't seem necessary, but again if someone wants to maintain all of this I'm not against having it.

sc250024 commented:
> That makes this look more official than it currently is, I think it should be jsonnet-bundler/jsonnet-bundler. As I said I don't feel too strongly, but docker containers don't seem necessary, but again if someone wants to maintain all of this I'm not against having it.

I still need to look into the Docker part of Drone CI, but assuming that it works, I would just need a decision as to where to put it.

sh0rez commented:
Hey, so afaict the last open question is the location of the docker hub image? I personally like `jsonnet/bundler`, because it would allow to co-colocate with other popular jsonnet tools like `jsonnet/fmt` or `jsonnet/jsonnet`.

[Citing a Jsonnet maintainer](https://github.com/google/jsonnet/issues/521#issuecomment-392218272), I think it's fair enough to do this:
> The jsonnet-bundler project is **almost-official**. It was consulted with @sparkprime and broader community and it has everyone's blessing. So we are counting on jsonnet-bundler becoming the standard for jsonnet dependency management. However, as you mentioned, it's still pretty alpha.

I can create the `jsonnet/bundler` repo if everyone is okay with this.

PS @sc250024: You seem to be force-pushing. This makes it slightly harder for us to review your PRs, because addressed review suggestions can't be viewed in their own commit, but must be identified in the big changeset, which is now different than it was before. It might be that you do this for having a clean commit history, but given we squash merge, this extra care is not neccessary :)

sc250024 commented:
> PS @sc250024: You seem to be force-pushing. This makes it slightly harder for us to review your PRs, because addressed review suggestions can't be viewed in their own commit, but must be identified in the big changeset, which is now different than it was before. It might be that you do this for having a clean commit history, but given we squash merge, this extra care is not neccessary :)

@sh0rez Ahh sorry, I'll stop doing that from now on. I just made the last force-push just now. Basically, I took your advice, and used Drone's native Docker building: http://plugins.drone.io/drone-plugins/drone-docker/

brancz commented:
I just want the jsonnet folks to be ok with it, and not invade their space. Obviously I'm happy with it being more official :)

sc250024 commented:
Updated docs to reflect Docker and Homebrew. Homebrew-core PR is here: https://github.com/Homebrew/homebrew-core/pull/58487

brancz commented:
Judging by demand in other projects, I expect that if we have containers, then people will soon ask for multi arch. Is there a good way that we can combine these concerns? With these changes we will already be building multi-arch.

sc250024 commented:
> Judging by demand in other projects, I expect that if we have containers, then people will soon ask for multi arch. Is there a good way that we can combine these concerns? With these changes we will already be building multi-arch.

I'll add them.

sc250024 commented:
@brancz I did some research, and while implementing multi-architecture Docker builds with Drone is possible, it means expanding the scope of this PR quite a bit.

See for yourself how Drone themselves, and Gitea implement multi-arch builds:

- **Drone**: Uses the Drone `manifest` plugin and multiple Dockerfiles
  - https://github.com/drone/drone/blob/master/.drone.yml
  - https://github.com/drone/drone/tree/master/docker
- **Gitea**:
  - https://github.com/go-gitea/gitea/blob/master/docker/manifest.tmpl
  - https://github.com/go-gitea/gitea/blob/master/Dockerfile
  - https://github.com/go-gitea/gitea/blob/master/docker/manifest.tmpl

This would mean scope-creeping a bit, and I would need more time to implement this since this *really* needs testing with Drone locally.

sh0rez commented:
We do this in Tanka and it works well.

Agree it's outside of this PR, let's keep this focused!

sc250024 commented:
> We do this in Tanka and it works well.
> 
> Agree it's outside of this PR, let's keep this focused!

If that's the case, and there's already an existing example that you all know works, I can open another PR after this to take care of the multi-arch build.

brancz commented:
That works for me. In that case I would prefer to keep all container images out of this PR and have the next PR do all the docker things, just so we don't end up in a somewhat "in progress" state.

sc250024 commented:
I'll make some of the "final" changes tonight. I'll also open an issue for multi-arch builds. It should be straightforward, but I would need to do some testing.

By the way, `jsonnet-bundler` is now in Homebrew => https://github.com/Homebrew/homebrew-core/pull/58487 🥳
